### PR TITLE
virtio: console and vsock support

### PR DESCRIPTION
**Pull request by andybui01**
This modifies the virtio console to support multiplexing, as well as add a new virtio vsock device.

### CODE REVIEW

File: libsel4vmmplatsupport/src/drivers/virtio_console_emul.c [hunk 1]
```diff
@@ -72,38 +77,122 @@ static void emul_con_rx_complete(void *iface, char *buf, unsigned int len)
...
+                con->con_count,
+                VIRTIO_CON_CON_PORT,
+                1
+            };
+            con->con_count++;
+            emul_con_rx_complete(emul, CTL_RX_QUEUE, (char *) &out_msg, sizeof(out_msg));
+        }
+        break;
+    default:
+        assert(!"panic");
```
[the remark this answers is not in the document]
Comment (axel-h) [reply libsel4vmmplatsupport/src/drivers/virtio_console_emul.c:175]: Printing an error with the event number would be helpful at least.

**Comment by andybui01** [comment]
Is this good to be merged? @axel-h @abrandnewusername

**Comment by andybui01** [comment]
Github does not let you request 2 reviews at once sorry for the notifications.

**Comment by lsf37** [comment]
> Github does not let you request 2 reviews at once sorry for the notifications.

No problem with the notification, just want to clarify that it does let your request multiple reviewers (just doesn't work in private repos without paid plans)

**Review by abrandnewusername — APPROVED** [review]

**Comment by andybui01** [comment]
Adding multiple console support adds another layer of indirection and will break the current Arm VMM (I'm trying to get the camkes-vm PR accepted at the same time), is everything all good otherwise?

**Comment by lsf37** [comment]
@abrandnewusername either this PR or seL4/seL4_projects_libs#74 is causing build failure in all projects that are using CAmkES currently (e.g https://github.com/seL4/camkes-vm-examples/actions/runs/3562188372). Can you have a look, please?

**Comment by lsf37** [comment]
Also, it looks like the PR removes old Copyright headers. That should not be done, Copyright lines should only be added, not removed.

**Comment by abrandnewusername** [comment]
> 

Do you mean for existed files, the Copyright headers should not be altered even the content of the file has been changed?

**Comment by abrandnewusername** [comment]
> @abrandnewusername either this PR or #74 is causing build failure in all projects that are using CAmkES currently (e.g https://github.com/seL4/camkes-vm-examples/actions/runs/3562188372). Can you have a look, please?

One of the commits in this [PR](https://github.com/seL4/camkes-vm/pull/36) should fix the compilation issue that causes the build failure, however, I don't think other commits in this PR are ready to be merged. Is it OK if I cherry-pick that one commit and make a pull request for it?

**Comment by lsf37** [comment]
> > Also, it looks like the PR removes old Copyright headers. That should not be done, Copyright lines should only be added, not removed.
> 
> Do you mean for existed files, the Copyright headers should not be altered even the content of the file has been changed?

Yes, the previous copyright holder does not lose their copyright when a file changes. In fact, they usually have copyright to the entire file still, including the new changes, because these tend to be derived work. But there can be additional copyright to the new additions to the file. This means, we should add copyright lines, but not remove them (unless there are special case like the entire file being removed and replaced with something else).

**Comment by lsf37** [comment]
> One of the commits in [this PR](https://github.com/seL4/camkes-vm/pull/36) should fix the compilation issue that causes the build failure, however, I don't think other commits in that PR are ready to be merged. Is it OK if I cherry-pick that one commit and make a pull request for it?

That sounds like a good strategy. @kent-mcleod were there other parts of this PR that were not ready yet?

**Comment by abrandnewusername** [comment]
> > > Also, it looks like the PR removes old Copyright headers. That should not be done, Copyright lines should only be added, not removed.
> > 
> > 
> > Do you mean for existed files, the Copyright headers should not be altered even the content of the file has been changed?
> 
> Yes, the previous copyright holder does not lose their copyright when a file changes. In fact, they usually have copyright to the entire file still, including the new changes, because these tend to be derived work. But there can be additional copyright to the new additions to the file. This means, we should add copyright lines, but not remove them (unless there are special case like the entire file being removed and replaced with something else).

What is the best way to fix this? Should I revert the PR or make another commit to fix Copyright headers?

**Comment by lsf37** [comment]
> What is the best way to fix this? Should I revert the PR or make another commit to fix Copyright headers?

A new commit to re-add them is probably easiest.

**Pull request merged**